packages/alfa-rules/src/sia-r7/rule.ts

```diff
@@ -46,22 +47,36 @@ export default Rule.Atomic.of<Page, Attribute>({
           if (test(and(isElement, hasAttribute("lang", not(isEmpty))), node)) {
             lang = node.attribute("lang");
           }
-
-          if (
-            test(
-              and(
-                isText,
+          if (lang.isSome()) {
+            if (
+              test(
+                and(
+                  isText,
+                  and(
+                    or(isVisible(device), not(isIgnored(device))),
+                    (text: Text) => !isWhitespace(text.data)
+                  )
+                ),
+                node
+              )
+            ) {
+              yield* lang;
+            }
+            if (
+              test(
                 and(
-                  or(isVisible(device), not(isIgnored(device))),
-                  (text: Text) => !isWhitespace(text.data)
-                )
-              ),
-              node
-            )
-          ) {
-            yield* lang;
+                  isElement,
+                  hasAccessibleName(
+                    device,
+                    (accessibleName) => !isWhitespace(accessibleName.value)
+                  )
+                ),
+                node
+              )
+            ) {
+              yield * lang;
+            }
```

Comment:
Could this be rewritten to something like this?
```JavaScript
const isVisibleText = test(
  and(
    isText,
    and(
      or(isVisible(device), not(isIgnored(device))),
      (text: Text) => !isWhitespace(text.data)
    )
  ),
  node
);

const isElementWithAccessibleName = const isElementWithAccessibleName = test(
  and(
    isElement,
    hasAccessibleName(
      device,
      (accessibleName) => !isWhitespace(accessibleName.value)
    )
  ),
  node
); 

if(isVisibleText || isElementWithAccessibleName) {
  yield * lang;
}
```

Reply:
Ah yes, that would be even more readable 👍 
We can go more "predicate" by doing
```javascript
const isVisibleText = and(
    isText,
    and(
      or(isVisible(device), not(isIgnored(device))),
      (text: Text) => !isWhitespace(text.data)
    )
  );

const isElementWithAccessibleName = and(
    isElement,
    hasAccessibleName(
      device,
      (accessibleName) => !isWhitespace(accessibleName.value)
    )
  ) 

if(test(or(isVisibleText, isElementWithAccessibleName), node)) {
  yield* lang;
}
```

Reply:
good spot!  I have changed that